File: problem102.py
# Definition for a binary tree nodes.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import deque
class Solution:
    def levelOrder(self, root):
        if not root:
            return []
        
        # Check left and right and pop it and add to answer
        q = deque() # We can check right side first and then left side
        ans = []
        if root:
            q.append(root)
            while q:
                length = len(q)
                nodes_vals = []
                for _ in range(length):
                    node = q.popleft() # get the first elemnt in q
                    nodes_vals.append(node.val)

                    # Check two sides
                    if node.left:
                        q.append(node.left)
                    if node.right:
                        q.append(node.right)
                ans.append(nodes_vals)
        return ans

File: problem102.cpp
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        // Write queue and ans vector
        queue<TreeNode*> q;
        vector<vector<int>> ans;
        if(!root){
            return ans; // base case 
        }
        q.push(root); // add the first node as root
        while(!q.empty()){
            int length = q.size();
            vector<int> node_vals;
            for(int i = 0;  i< length; i++){
                TreeNode* node = q.front();
                node_vals.push_back(node -> val);
                q.pop();
                if(node -> left){
                    q.push(node -> left);
                }
                if(node -> right){
                    q.push(node -> right);
                }
            }
            ans.push_back(node_vals);
        }
        return ans;
    }
};